Style evolution (#3338)
rust-lang/rfcs: text/3338-style-evolution.md

style. The default Rust style varies by Rust edition. (Not every edition
changes the Rust style, and thus some editions have identical default styles;
Rust 2015, 2018, and 2021 all have the same default style.)

Direct invocations of `rustfmt` obtain the edition used for parsing Rust code
from the `edition` option in its configuration file (`rustfmt.toml` or
`.rustfmt.toml`), or via the `--edition` command-line option; `cargo fmt`
obtains the edition from the `edition` option in `Cargo.toml` and passes it to
`rustfmt`. By default, `rustfmt` and `cargo fmt` will use the same edition for
style as the Rust edition used for parsing.

However, when transitioning between editions, projects may want to separately
make and commit the changes for 1) transitioning to a new Rust edition and 2)
transitioning to a new style edition. To allow for this, `rustfmt` also allows
configuring the style edition directly, via a separate `style_edition`
configuration option, or `--style-edition` command-line option. `style_edition`
or `--style-edition`, if set, always overrides `edition` or `--edition` for the
purposes of styling, though `edition` or `--edition` still determines the

> edition for the purposes of parsing Rust code.

Reviewer: Though perhaps implicit, I think the only piece not explicitly acknowledged in this is that some projects may not want to change their style edition/introduce formatting changes whatsoever (the "I want to update to the next `edition` but i don't care about/want any formatting changes" crowd).

Do you think that's worth calling out explicitly? Otherwise one could reasonably interpret (at least on a first pass) this as forcing formatting changes, but in a sequenced manner

Author: @calebcartwright I think it's worth discussing this explicitly; I've added some language discussing this.

from the `edition` option in its configuration file (`rustfmt.toml` or
`.rustfmt.toml`), or via the `--edition` command-line option; `cargo fmt`
obtains the edition from the `edition` option in `Cargo.toml` and passes it to
`rustfmt`. By default, `rustfmt` and `cargo fmt` will use the same edition for
style as the Rust edition used for parsing.

However, when transitioning between editions, projects may want to separately
make and commit the changes for 1) transitioning to a new Rust edition and 2)
transitioning to a new style edition. To allow for this, `rustfmt` also allows
configuring the style edition directly, via a separate `style_edition`
configuration option, or `--style-edition` command-line option. `style_edition`
or `--style-edition`, if set, always overrides `edition` or `--edition` for the
purposes of styling, though `edition` or `--edition` still determines the
edition for the purposes of parsing Rust code.

Note that rustfmt may not necessarily support all combinations of Rust edition
and style edition; in particular, it may not support using a style edition that
differs by more than one step from the Rust edition.

> The current version of the style guide will describe the latest Rust edition.
> Each distinct past style will have a corresponding archived version of the
> style guide. Note that archived versions of the style guide may not necessarily
> document formatting for newer Rust constructs that did not exist at the time
> that version of the style guide was archived.

Reviewer 2: objection: I'm not confident that this will end up being the best way to track the different style editions and I'd prefer not to specify the specific mechanism as part of this RFC and rather wait until we've tried splitting the style guide for two editions to see what approach we end up with.

```suggestion
The style team will maintain distinct versions of the style guide for each style
edition. Note that archived versions of the style guide may not necessarily
document formatting for newer Rust constructs that did not exist at the time
that version of the style guide was archived.
```
